<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<?xml-stylesheet type='text/xsl' href='options.xsl'?><html xmlns="http://www.w3.org/1999/xhtml">
  <head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <title>mcast - a motif search tool</title>
    <link href="./meme-suite.css" rel="stylesheet" type="text/css" />
  </head>
  <body>
    <div id="main">
      <script src="./meme-suite-logo.js" type="text/javascript">&gt;</script>
      <img src="images/mcast-logo.png" alt="MCAST logo" />
      <p><b>Usage:</b></p>
      <code>mcast [options] &lt;motifs&gt; &lt;database&gt;</code>
      <p><b>Description:</b></p>
      <p>
      <code>MCAST</code> searches a sequence database for
      statistically significant clusters of non-overlapping "hits" to the
      motifs in a query.</p>
      <p>A "hit" is a sequence position that is sufficiently similar to a
      motif in the query. To be a hit, the <i>p</i>-value of the motif
      alignment score must be less than the significance threshold,
      <i>pthresh</i> (see option --p-thresh, below). The alignment of the motif
      and the sequence position is done <b>without</b> gaps. To compute
      the <i>p-</i>value of a motif alignment score, <code>MCAST</code>
      assumes that the sequences in the database were generated by a
      0-order Markov process; see option --bgfile, below. With DNA sequences,
      <code>MCAST</code> searches for hits on both the sequences given in
      the database, and their reverse complements.</p>
      <p>A cluster of non-overlapping hits is called a "match". The user
      specifies the maximum allowed distance between the hits in a match
      using the --max-gap option.
      (Two hits separated by more than the maximum allowed gap will be
      reported in separate matches.)</p>
      <p><code>MCAST</code> searches for all of the matches between the
      query and the sequences in the database. Each match is assigned an
      <i>E</i>-value, and matches that score below an <i>E</i>-value
      threshold are printed in order of increasing <i>E</i>-value (see
      option --e-thresh, below).</p>
      <p>The <i>p-</i>value of a hit is converted to a "p-score" in order
      to compute the total score of the match it participates in. 
      The p-score for a hit with <i>p-</i>value <i>p</i> is</p>
      <div style="margin-left: 2em"><p><i>S =
      -log<sub>2</sub>(p/pthresh),</i></p></div>
      <p>
      where the significance threshold <i>pthresh</i> may be specified by
      the user. The total score of a match is the sum of the p-scores of
      the hits making up the match. <code>MCAST</code> finds the matches
      with the maximum match scores.
      </p>
      <p>In order for <i>E</i>-values to be computed by
      <code>MCAST</code>, at least 100 matches must be found. If there
      are too few sequences in the database, or if certain other options
      are made too stringent (see Options, below), too few matches may
      exist for <i>E</i>-values to be computed. In this case, the results
      are sorted by match score, the <i>E</i>-value column is set to
      "NaN" and all matches are printed. 
      This limitation can be overcome by specifying the --synth and --bgfile options. 
      When those options are set, synthetic sequences will
      be generated from the provided background model, and used to estimate
      <i>E</i>-values.</p>
      <p>
      A full description of the algorithm is found in:
      </p>
      <blockquote>
      <p>
      <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/suppl_2/ii16">
      Bailey and Noble. "Searching for statistically significant regulatory
      modules." <i>Bioinformatics (Proceedings of the European Conference on
      Computational Biology)</i>. 19(Suppl. 2):ii16-ii25, 2003.</a>
      </p>
      </blockquote>
      <p><b>Input:</b></p>
      <ul> 
      <li>
      <code>&lt;motifs&gt;</code> is a list of motifs, in <A HREF="meme-format.html">MEME format</A>.
      </li>
      <li>
      <code>&lt;database&gt;</code> is a collection of sequences in
      <A HREF="fasta-format.html">FASTA</A> format. 
      </li>
      </ul>
      <p><b>Output:</b></p>
      <ul>
      <li>An HTML file containing a table listing the high scoring matches
      (<a href="examples/mcast_example_output_files/mcast.html">sample</a>).
      </li>
      <!--
      An XML file using the 
      <a href="http://zlab.bu.edu/~phaverty/SupplementalData/CisML">CisML</a>
      schema.
      -->
      </ul>
      <p><b>Options:</b></p>
      <ul>
      <li><code>--bgfile <i>&lt;bfile&gt;</i></code> 
    - Read background frequencies from
    <code><i>&lt;bfile&gt;</i></code>.  
    The file should be in 
    <a href="bfile-format.html">MEME background file format</a>.
    The default is to use frequencies
    embedded in the application from the non-redundant database.  If the
    argument is the keyword <code>motif-file</code>, then the frequencies
    will be taken from the motif file.
  </li>
      <li><code>--bgweight <i>&lt;weight&gt;</i></code> 
    - Add <i>&lt;weight&gt;</i> times the background frequency to
    the corresponding letter counts in each motif when
    converting them to postion specific scoring matrices.
    The default value is 4.0.
  </li>
      <li><code>--e-thresh <i>&lt;ev&gt;</i></code>
    - Only print results with E-values less than 
    <code><i>&lt;ev&gt;</i></code>.
    The default threshold is 10.0.
  </li>
      <li><code>--max-gap <i>&lt;max-gap&gt;</i></code>
    - The value of <code><i>&lt;max-gap&gt;</i></code> specifies
    the longest distance allowed between two hits in a match.
    Hits separated by more than <code><i>&lt;max-gap&gt;</i></code> 
    will be placed in different matches. The default value is 50.
    <b>Note</b>:
    Large values of <code><i>&lt;max-gap&gt;</i></code> combined with
    large values of <i>pthresh</i> may prevent <code>MCAST</code> from
    computing <i>E</i>-values.
  </li>
      <li><code>--o <i>&lt;dir name&gt;</i></code>
    - Specifies the output directory.
    If the directory already exists, the contents will not be overwritten.
  </li>
      <li><code>--oc <i>&lt;dir name&gt;</i></code>
    - Specifies the output directory.
    If the directory already exists, the contents will be overwritten.
  </li>
      <li><code>--p-thresh <i>&lt;pv&gt;</i></code>
    - Only motif occurences with p-values less than 
    <code><i>&lt;pv&gt;</i></code> will be considered in computing
    the match score. 
    The default value is 5e-4.
  </li>
      <li><code>--synth</code>
    - Create synthetic sequences for estimating
    <i>E</i>-values. This is useful with small input databases where
    not enough match scores are found to estimate <i>E</i>-values.
    The <code>--bgfile</code> option must also be set when
    using this option.
  </li>
      <li><code>--text</code>
    Output is plain text rather then HTML.
  </li>
      <li><code>--transfac</code>
    - The input motif file is assumed to be in 
    <a href="transfac-format.html">TRANSFAC format</a>
    and is converted to 
    <a href="meme-format.html">MEME format</a>
    before being used.
  </li>
      <li><code>--verbosity 1|2|3|4</code>
    - Set the verbosity of status reports to standard error.
    The default level is 2.
  </li>
      </ul>
    </div>
  </body>
</html>
